# 📊 Loan Approval Prediction System
## Step 1: Exploratory Data Analysis & Inspection

Welcome to the data exploration notebook for the **Loan Approval Prediction System**.
In this step, our goal is to understand the dataset structure, inspect the rows and columns, check for data cleanliness (missing values, duplicates), inspect the target variable, and understand the meaning of each feature.

---


In [1]:
import os
import pandas as pd

print(f"Pandas version: {pd.__version__}")

Pandas version: 2.2.3


### 📁 Loading the Dataset
We load `loan_approval_dataset.csv` from the `../data/` directory.

> **Note for Beginners:** In raw datasets, column names and text entries often have accidental leading or trailing spaces. We remove these whitespaces using `.str.strip()` so columns can be indexed cleanly without issues later.


In [2]:
# Load CSV file
data_path = os.path.join('..', 'data', 'loan_approval_dataset.csv')
df = pd.read_csv(data_path)

# Clean column headers (strip leading/trailing whitespaces)
df.columns = df.columns.str.strip()

# Clean categorical string columns (strip leading/trailing whitespaces)
for col in df.select_dtypes(include=['object']).columns:
    df[col] = df[col].str.strip()

print("Dataset loaded and cleaned successfully!")

Dataset loaded and cleaned successfully!


### 1 & 2. Number of Rows and Columns

In [3]:
num_rows, num_columns = df.shape
print(f"1. Total Number of Rows    : {num_rows:,}")
print(f"2. Total Number of Columns : {num_columns}")

1. Total Number of Rows    : 4,269
2. Total Number of Columns : 13


### 3. All Column Names

In [4]:
print("All Column Names:")
for i, col in enumerate(df.columns, 1):
    print(f"  {i:2d}. {col}")

All Column Names:
   1. loan_id
   2. no_of_dependents
   3. education
   4. self_employed
   5. income_annum
   6. loan_amount
   7. loan_term
   8. cibil_score
   9. residential_assets_value
  10. commercial_assets_value
  11. luxury_assets_value
  12. bank_asset_value
  13. loan_status


### 4. First 5 Rows
Let's preview the first 5 records of the dataset.

In [5]:
df.head()

,loan_id,no_of_dependents,education,self_employed,income_annum,loan_amount,loan_term,cibil_score,residential_assets_value,commercial_assets_value,luxury_assets_value,bank_asset_value,loan_status
0,1,2,Graduate,No,9600000,29900000,12,778,2400000,17600000,22700000,8000000,Approved
1,2,0,Not Graduate,Yes,4100000,12200000,8,417,2700000,2200000,8800000,3300000,Rejected
2,3,3,Graduate,No,9100000,29700000,20,506,7100000,4500000,33300000,12800000,Rejected
3,4,3,Graduate,No,8200000,30700000,8,467,18200000,3300000,23300000,7900000,Rejected
4,5,5,Not Graduate,Yes,9800000,24200000,20,382,12400000,8200000,29400000,5000000,Rejected


### 5. Data Types
Understanding the data type (`int64`, `float64`, or `object` for strings) tells us which features are numerical and which are categorical.

In [6]:
print("Data types of each column:")
print(df.dtypes)

Data types of each column:
loan_id                      int64
no_of_dependents             int64
education                   object
self_employed               object
income_annum                 int64
loan_amount                  int64
loan_term                    int64
cibil_score                  int64
residential_assets_value     int64
commercial_assets_value      int64
luxury_assets_value          int64
bank_asset_value             int64
loan_status                 object
dtype: object


### 6. Missing Values
Checking if any fields have null or missing (`NaN`) values.

In [7]:
missing_counts = df.isnull().sum()
print("Missing values per column:")
print(missing_counts)
print(f"\nTotal missing values in entire dataset: {missing_counts.sum()}")

Missing values per column:
loan_id                     0
no_of_dependents            0
education                   0
self_employed               0
income_annum                0
loan_amount                 0
loan_term                   0
cibil_score                 0
residential_assets_value    0
commercial_assets_value     0
luxury_assets_value         0
bank_asset_value            0
loan_status                 0
dtype: int64

Total missing values in entire dataset: 0


### 7. Duplicate Rows
Checking if any identical applicant records appear multiple times.

In [8]:
duplicate_count = df.duplicated().sum()
print(f"Total duplicate rows: {duplicate_count}")

Total duplicate rows: 0


### 8. Unique Values of Categorical Columns
Checking the distinct options/categories present in each text column.

In [9]:
cat_cols = df.select_dtypes(include=['object']).columns.tolist()
print("Categorical columns and their unique values:\n")
for col in cat_cols:
    uniques = df[col].unique().tolist()
    print(f"- Column: '{col}'")
    print(f"  Unique Values ({len(uniques)}): {uniques}\n")

Categorical columns and their unique values:

- Column: 'education'
  Unique Values (2): ['Graduate', 'Not Graduate']

- Column: 'self_employed'
  Unique Values (2): ['No', 'Yes']

- Column: 'loan_status'
  Unique Values (2): ['Approved', 'Rejected']



### 9 & 10. Target Column and Value Distribution
The target column is **`loan_status`**, which indicates whether a loan was approved or rejected.

In [10]:
target_col = 'loan_status'
print(f"Target Column: '{target_col}'\n")

counts = df[target_col].value_counts()
proportions = df[target_col].value_counts(normalize=True) * 100

summary_df = pd.DataFrame({
    'Count': counts,
    'Percentage (%)': proportions.round(2)
})
print("Target Value Distribution:")
print(summary_df)

Target Column: 'loan_status'

Target Value Distribution:
             Count  Percentage (%)
loan_status                       
Approved      2656           62.22
Rejected      1613           37.78


---
## 📖 Column Explanations (Simple Language for Beginners)

Here is what each column in this dataset means:

| Column Name | Data Type | Description |
|:---|:---|:---|
| **`loan_id`** | Numerical (int) | Unique identification number assigned to each loan applicant (not a predictive feature). |
| **`no_of_dependents`** | Numerical (int) | Number of family members financially dependent on the applicant. |
| **`education`** | Categorical (string) | Educational qualification of the applicant (`Graduate` or `Not Graduate`). |
| **`self_employed`** | Categorical (string) | Employment type (`Yes` if running own business/freelance, `No` if salaried/employed). |
| **`income_annum`** | Numerical (int) | Annual total earnings of the applicant (in Indian Rupees ₹). |
| **`loan_amount`** | Numerical (int) | Total amount of loan requested by the applicant (in ₹). |
| **`loan_term`** | Numerical (int) | Loan repayment period in years (e.g., 2, 4, 10, 20 years). |
| **`cibil_score`** | Numerical (int) | Credit score measuring creditworthiness (ranges from 300 to 900). Higher scores mean better repayment history. |
| **`residential_assets_value`** | Numerical (int) | Total market value of residential homes/properties owned by the applicant. |
| **`commercial_assets_value`** | Numerical (int) | Total market value of commercial properties/shops owned by the applicant. |
| **`luxury_assets_value`** | Numerical (int) | Total value of luxury possessions owned (cars, jewelry, paintings, etc.). |
| **`bank_asset_value`** | Numerical (int) | Liquid cash and balance held across the applicant's bank accounts. |
| **`loan_status`** *(Target)* | Categorical (string) | The outcome to predict: **`Approved`** or **`Rejected`**. |

---
## 💡 Summary of Initial Findings
1. **Clean Data**: There are **0 missing values** and **0 duplicate rows** in the entire dataset.
2. **Dataset Size**: We have **4,269 applicants** across **13 attributes** (1 ID + 11 input features + 1 target).
3. **Class Distribution**:
   - **Approved**: 2,656 (~62.2%)
   - **Rejected**: 1,613 (~37.8%)
   - Moderate balance (no extreme class imbalance).
4. **Next Steps Ahead**: In subsequent steps, we can perform feature engineering, encode categorical features, scale numerical values, and train classification models.
